prog/encodingexec.go

```diff
@@ -49,13 +79,27 @@ func (p *Prog) SerializeForExec(buffer []byte, pid int) error {
 	for _, c := range p.Calls {
 		// Calculate checksums.
 		csumMap := calcChecksumsCall(c, pid)
+		csumUses := make(map[*Arg]bool)
+		for arg, info := range csumMap {
+			csumUses[arg] = true
+			if info.Kind == CsumInet {
+				for _, chunk := range info.Chunks {
+					if chunk.Kind == CsumChunkArg {
+						csumUses[chunk.Arg] = true
+					}
+				}
+			}
+		}
 		// Calculate arg offsets within structs.
 		// Generate copyin instructions that fill in data into pointer arguments.
 		foreachArg(c, func(arg, _ *Arg, _ *[]*Arg) {
 			if arg.Kind == ArgPointer && arg.Res != nil {
 				foreachSubargOffset(arg.Res, func(arg1 *Arg, offset uintptr) {
-					if len(arg1.Uses) != 0 {
-						w.args[arg1] = argInfo{Offset: offset}
+					if len(arg1.Uses) != 0 || csumUses[arg1] {
+						w.args[arg1] = argInfo{Addr: physicalAddr(arg) + offset}
+					}
+					if arg1.Kind == ArgGroup || arg1.Kind == ArgUnion {
```

Comment:
Can ArgGroup/ArgUnion have len(arg1.Uses) != 0 || csumUses[arg1]?
I think we only use/reference scalars? If so, I don't understand why you changed foreachSubargOffset to pass ArgGroup/ArgUnion to the callback and then filter them out here.